src/memray/_memray/source.cpp

```diff
@@ -63,6 +75,21 @@ FileSource::close()
     _close();
 }
 
+void
+FileSource::findReadableSize()
+{
+    d_raw_stream->seekg(-1, d_raw_stream->end);
+    while (*d_raw_stream) {
```

Comment:
I know it can't happen, but we may want to check also that we don't go past the beginning of the file.

Reply:
I just checked the behavior - if we try, `*d_raw_stream` will be false, because `failbit` gets set, so we break out of the loop without changing `d_readable_size` from its default value of `0` - so all future `read()` calls will fail.

And, the only way this can happen is if someone overwrites the file with all zeroes while we were in the middle of reading it, since we've already read some non-zero bytes by this point. I don't think this is worth worrying about.

Reply:
I'll add a comment explaining what will happen, though:
```c
        // If we're at BOF, this sets failbit and makes the loop break.
        d_raw_stream->seekg(-1, d_raw_stream->cur);
```